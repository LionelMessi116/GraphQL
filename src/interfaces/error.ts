export interface IGraphQLError {
  message?: string;
}
