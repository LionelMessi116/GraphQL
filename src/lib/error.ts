import { GraphQLError } from '../types';

const generateErrorMessage = (
  networkErr?: Error,
  graphQlErrs?: GraphQLError[]
) => {
  let error = '';
  if (networkErr) {
    error = `[Network] ${networkErr.message}`;
    return error;
  }

  graphQlErrs.forEach(err => {
    error += `[GraphQL] ${err.message}\n`;
  });

  return error.trim();
};

const rehydrateGraphQlError = (error: string | GraphQLError): Error => {
  if (typeof error === 'string') {
    return new Error(error);
  } else if (error.message) {
    return new Error(error.message);
  } else {
    return error as any;
  }
};

/** An error which can consist of GraphQL errors and Network errors. */
export class CombinedError implements Error {
  public name: string;
  public message: string;
  public graphQLErrors: Error[];
  public networkError?: Error;
  public response?: any;

  constructor({
    networkError,
    graphQLErrors,
    response,
  }: {
    networkError?: Error;
    graphQLErrors?: Array<string | GraphQLError>;
    response?: any;
  }) {
    this.name = 'CombinedError';
    this.graphQLErrors = (graphQLErrors || []).map(rehydrateGraphQlError);
    this.message = generateErrorMessage(networkError, this.graphQLErrors);
    this.networkError = networkError;
    this.response = response;
  }
}
