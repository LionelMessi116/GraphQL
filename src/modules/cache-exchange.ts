import Observable from 'zen-observable-ts';

import { ICache, IExchange, IExchangeResult } from '../interfaces/index';
import { gankTypeNamesFromResponse } from './typenames';

// Wraps an exchange and refers/updates the cache according to operations
export const cacheExchange = (cache: ICache, forward: IExchange): IExchange => {
  return operation => {
    const withTypenames$ = forward(operation).map(
      (response: IExchangeResult) => {
        // Grab typenames from response data
        const typeNames = gankTypeNamesFromResponse(response.data);
        return { ...response, typeNames };
      }
    );

    const { operationName, key, context } = operation;
    if (operationName !== 'query') {
      return withTypenames$;
    }

    const { skipCache = false } = context;

    return new Observable<IExchangeResult>(observer => {
      let subscription;

      cache.read(key).then(cachedResult => {
        if (cachedResult && !skipCache) {
          observer.next(cachedResult);
          observer.complete();
          return;
        }

        subscription = withTypenames$.subscribe({
          complete: () => observer.complete(),
          error: err => observer.error(err),
          next: (response: IExchangeResult) => {
            // Store data in cache, using serialized query as key
            cache.write(key, response);
            // Return response with typeNames
            observer.next(response);
          },
        });
      });

      return () => {
        if (subscription) {
          subscription.unsubscribe();
        }
      };
    });
  };
};
