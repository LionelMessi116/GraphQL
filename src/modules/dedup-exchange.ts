import Observable from 'zen-observable-ts';

import { IExchange } from '../interfaces/exchange';

// Wraps an exchange and deduplicates in-flight operations by their key
export const dedupExchange = (forward: IExchange): IExchange => {
  const inFlight = {};

  return operation => {
    const { key } = operation;

    // Take existing intermediate observable if it has been created
    if (inFlight[key] !== undefined) {
      return inFlight[key];
    }

    // Keep around one subscription and collect observers for this observable
    const observers = [];
    let refCounter = 0;
    let subscription;

    // Create intermediate observable and only forward to the next exchange once
    return (inFlight[key] = new Observable(observer => {
      refCounter++;
      observers.push(observer);

      if (subscription === undefined) {
        subscription = forward(operation).subscribe({
          complete: () => {
            observers.forEach(x => x.complete());
            delete inFlight[key];
          },
          error: error => {
            observers.forEach(x => x.error(error));
          },
          next: emission => {
            observers.forEach(x => x.next(emission));
          },
        });
      }

      return () => {
        refCounter--;
        if (refCounter === 0) {
          delete inFlight[key];
          if (subscription !== undefined) {
            subscription.unsubscribe();
          }
        }
      };
    }));
  };
};
