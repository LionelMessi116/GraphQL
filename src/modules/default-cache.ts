import { ICache } from '../interfaces/index';

export const defaultCache = (store: object): ICache => ({
  invalidate: hash =>
    new Promise(resolve => {
      delete store[hash];
      resolve(hash);
    }),
  invalidateAll: () =>
    new Promise(resolve => {
      store = {};
      resolve();
    }),
  read: hash =>
    new Promise(resolve => {
      resolve(store[hash] || null);
    }),
  update: callback =>
    new Promise(resolve => {
      if (typeof callback === 'function') {
        Object.keys(store).forEach(key => {
          callback(store, key, store[key]);
        });
      }
      resolve();
    }),
  write: (hash, data) =>
    new Promise(resolve => {
      store[hash] = data;
      resolve(hash);
    }),
});
