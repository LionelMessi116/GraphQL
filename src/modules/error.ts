import { IGraphQLError } from '../interfaces/index';

const generateErrorMessage = (
  networkErr?: Error,
  graphQlErrs?: IGraphQLError[]
) => {
  let error = '';
  if (networkErr) {
    error = `[Network] ${networkErr.message}`;
    return error;
  }

  graphQlErrs.forEach(err => {
    error += `[GraphQL] ${err.message}\n`;
  });

  return error;
};

const rehydrateGraphQlError = (error: string | IGraphQLError): Error => {
  if (typeof error === 'string') {
    return new Error(error);
  } else if (error.message) {
    return new Error(error.message);
  } else {
    return error as any;
  }
};

export class CombinedError extends Error {
  public message: string;
  public graphQLErrors: Error[];
  public networkError?: Error;

  constructor({
    networkError,
    graphQLErrors,
  }: {
    networkError?: Error;
    graphQLErrors?: Array<string | IGraphQLError>;
  }) {
    const gqlErrors = (graphQLErrors || []).map(rehydrateGraphQlError);
    const errorMessage = generateErrorMessage(networkError, gqlErrors);

    super(errorMessage);

    this.message = errorMessage;
    this.graphQLErrors = gqlErrors;
    this.networkError = networkError;

    Object.setPrototypeOf(this, CombinedError.prototype);
  }
}
