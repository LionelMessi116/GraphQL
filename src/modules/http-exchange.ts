import Observable from 'zen-observable-ts';

import { IExchange } from '../interfaces/index';
import { CombinedError } from './error';

const checkStatus = (redirectMode: string = 'follow') => (
  response: Response
) => {
  // If using manual redirect mode, don't error on redirect!
  const statusRangeEnd = redirectMode === 'manual' ? 400 : 300;
  if (response.status >= 200 && response.status < statusRangeEnd) {
    return response;
  }

  throw new Error(response.statusText);
};

const createAbortController = () => {
  if (typeof AbortController === 'undefined') {
    return { abort: null, signal: undefined };
  }

  return new AbortController();
};

export const httpExchange = (): IExchange => operation => {
  const { url, fetchOptions } = operation.context;

  const body = JSON.stringify({
    query: operation.query,
    variables: operation.variables,
  });

  // https://developer.mozilla.org/en-US/docs/Web/API/AbortController/AbortController
  const abortController = createAbortController();

  return new Observable(observer => {
    let response;

    fetch(url, {
      body,
      headers: { 'Content-Type': 'application/json' },
      method: 'POST',
      signal: abortController.signal,
      ...fetchOptions,
    })
      .then(res => (response = res))
      .then(checkStatus(fetchOptions.redirect))
      .then(res => res.json())
      .then(result => {
        let error;
        if (Array.isArray(result.errors)) {
          error = new CombinedError({
            graphQLErrors: result.errors,
            response,
          });
        }

        if (result.data) {
          observer.next({
            data: result.data,
            error,
          });
          observer.complete();
        } else if (error) {
          observer.error(error);
        } else {
          observer.error(new Error('no data or error'));
        }
      })
      .catch(err => {
        if (err.name === 'AbortError') {
          return;
        }

        const error = new CombinedError({
          networkError: err,
          response,
        });

        observer.error(error);
      });

    return () => {
      if (abortController.abort) {
        abortController.abort();
      }
    };
  });
};
