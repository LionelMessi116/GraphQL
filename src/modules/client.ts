import Observable from 'zen-observable-ts';

import {
  ICache,
  IClientOptions,
  IExchange,
  IExchangeResult,
  IQuery,
} from '../interfaces/index';

import { cacheExchange } from './cache-exchange';
import { dedupExchange } from './dedup-exchange';
import { defaultCache } from './default-cache';
import { hashString } from './hash';
import { httpExchange } from './http-exchange';

export default class Client {
  url: string;
  store: object; // Internal store
  subscriptions: object; // Map of subscribed Connect components
  subscriptionSize: number; // Used to generate IDs for subscriptions
  cache: ICache; // Cache object
  exchange: IExchange; // Exchange to communicate with GraphQL APIs
  fetchOptions: RequestInit | (() => RequestInit); // Options for fetch call

  constructor(opts?: IClientOptions) {
    if (!opts) {
      throw new Error('Please provide configuration object');
    } else if (!opts.url) {
      throw new Error('Please provide a URL for your GraphQL API');
    }

    this.url = opts.url;
    this.store = opts.initialCache || {};
    this.subscriptions = {};
    this.subscriptionSize = 0;
    this.cache = opts.cache || defaultCache(this.store);
    this.exchange = cacheExchange(this.cache, dedupExchange(httpExchange()));
    this.fetchOptions = opts.fetchOptions || {};

    // Bind methods
    this.executeQuery = this.executeQuery.bind(this);
    this.executeMutation = this.executeMutation.bind(this);
    this.updateSubscribers = this.updateSubscribers.bind(this);
    this.refreshAllFromCache = this.refreshAllFromCache.bind(this);
  }

  updateSubscribers(typenames, changes) {
    // On mutation, call subscribed callbacks with eligible typenames
    for (const sub in this.subscriptions) {
      if (this.subscriptions.hasOwnProperty(sub)) {
        this.subscriptions[sub](typenames, changes);
      }
    }
  }

  subscribe(
    callback: (changedTypes: string[], response: object) => void
  ): () => void {
    // Create an identifier, add callback to subscriptions
    const id = this.subscriptionSize++;
    this.subscriptions[id] = callback;

    // Return unsubscription function
    return () => {
      delete this.subscriptions[id];
    };
  }

  refreshAllFromCache() {
    // On mutation, call subscribed callbacks with eligible typenames
    return new Promise(resolve => {
      for (const sub in this.subscriptions) {
        if (this.subscriptions.hasOwnProperty(sub)) {
          this.subscriptions[sub](null, null, true);
        }
      }
      resolve();
    });
  }

  makeContext({ skipCache }: { skipCache?: boolean }): Record<string, any> {
    return {
      fetchOptions:
        typeof this.fetchOptions === 'function'
          ? this.fetchOptions()
          : this.fetchOptions,
      skipCache: !!skipCache,
      url: this.url,
    };
  }

  executeQuery$(
    queryObject: IQuery,
    skipCache: boolean
  ): Observable<IExchangeResult> {
    // Create hash key for unique query/variables
    const { query, variables } = queryObject;
    const key = hashString(JSON.stringify({ query, variables }));
    const operation = {
      context: this.makeContext({ skipCache }),
      key,
      operationName: 'query',
      query,
      variables,
    };

    return this.exchange(operation);
  }

  executeQuery(
    queryObject: IQuery,
    skipCache: boolean
  ): Promise<IExchangeResult> {
    return new Promise<IExchangeResult>((resolve, reject) => {
      this.executeQuery$(queryObject, skipCache).subscribe({
        error: reject,
        next: resolve,
      });
    });
  }

  executeMutation$(
    mutationObject: IQuery
  ): Observable<IExchangeResult['data']> {
    // Create hash key for unique query/variables
    const { query, variables } = mutationObject;
    const key = hashString(JSON.stringify({ query, variables }));

    const operation = {
      context: this.makeContext({}),
      key,
      operationName: 'mutation',
      query,
      variables,
    };

    return this.exchange(operation).map((response: IExchangeResult) => {
      // Notify subscribed Connect wrappers
      this.updateSubscribers(response.typeNames, response);
      // Resolve result
      return response.data;
    });
  }

  executeMutation(mutationObject: IQuery): Promise<IExchangeResult['data']> {
    return new Promise<IExchangeResult>((resolve, reject) => {
      this.executeMutation$(mutationObject).subscribe({
        error: reject,
        next: resolve,
      });
    });
  }
}
