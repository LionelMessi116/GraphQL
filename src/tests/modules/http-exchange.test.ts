import { httpExchange } from '../../modules/http-exchange';
import { CombinedError } from '../../modules/error';

describe('httpExchange', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    jest.clearAllMocks();
    jest.restoreAllMocks();
    (global as any).AbortController = (global as any)._AbortController;
  });

  afterEach(() => {
    (global as any).AbortController = undefined;
  });

  it('accepts an operation and sends a request', done => {
    (global as any).fetch.mockReturnValue(
      Promise.resolve({
        json: () => ({ data: [{ id: 5 }] }),
        status: 200,
      })
    );

    const exchange = httpExchange();
    const operation = {
      context: {
        fetchOptions: {
          test: 1,
        },
        url: 'http://localhost:3000/graphql',
      },
      key: 'test',
      operationName: 'query',
      query: `{ ping }`,
      variables: {},
    };

    exchange(operation).subscribe({
      error: err => {
        throw err;
      },
      next: x => {
        expect(x).toEqual({ data: [{ id: 5 }] });
        expect((global as any).fetch).toHaveBeenCalledWith(
          'http://localhost:3000/graphql',
          {
            body: JSON.stringify({
              query: `{ ping }`,
              variables: {},
            }),
            headers: { 'Content-Type': 'application/json' },
            method: 'POST',
            signal: expect.any(Object),
            test: 1,
          }
        );

        done();
      },
    });
  });

  it('errors with CombinedError', done => {
    (global as any).fetch.mockReturnValue(
      Promise.resolve({
        json: () => ({ errors: ['error message'] }),
        status: 200,
      })
    );

    const exchange = httpExchange();
    const operation = {
      context: {
        fetchOptions: {},
        url: 'http://localhost:3000/graphql',
      },
      key: 'test',
      operationName: 'query',
      query: `{ ping }`,
    };

    exchange(operation).subscribe({
      error: err => {
        expect(err).toBeInstanceOf(CombinedError);
        done();
      },
      next: () => {
        throw new Error('Should not be called');
      },
    });
  });

  it('is cancellable using an unsubscription', () => {
    const abort = jest.fn();
    const signal = { signal: 1 };

    (global as any).AbortController = jest.fn().mockImplementation(() => ({
      abort,
      signal,
    }));

    (global as any).fetch.mockReturnValue(
      new Promise(resolve => {
        setTimeout(
          () =>
            resolve({
              json: () => ({ errors: ['should not be called'] }),
              status: 200,
            }),
          50
        );
      })
    );

    const exchange = httpExchange();
    const operation = {
      context: {
        fetchOptions: {},
        url: 'http://localhost:3000/graphql',
      },
      key: 'test',
      operationName: 'query',
      query: `{ ping }`,
    };

    const subscription = exchange(operation).subscribe({
      error: () => {
        throw new Error('Should not be called');
      },
      next: () => {
        throw new Error('Should not be called');
      },
    });

    subscription.unsubscribe();

    expect((global as any).fetch).toHaveBeenCalledWith(
      'http://localhost:3000/graphql',
      {
        body: JSON.stringify({ query: `{ ping }` }),
        headers: { 'Content-Type': 'application/json' },
        method: 'POST',
        signal,
      }
    );

    expect(abort).toHaveBeenCalled();
  });

  it('throws when no data or errors have been returned', done => {
    (global as any).fetch.mockReturnValue(
      Promise.resolve({
        json: () => ({}),
        status: 200,
      })
    );

    const exchange = httpExchange();
    const operation = {
      context: {
        fetchOptions: {},
        url: 'http://localhost:3000/graphql',
      },
      key: 'test',
      operationName: 'query',
      query: `{ ping }`,
    };

    exchange(operation).subscribe({
      error: err => {
        expect(err.message).toBe('no data or error');
        done();
      },
      next: () => {
        throw new Error('Should not be called');
      },
    });
  });
});
