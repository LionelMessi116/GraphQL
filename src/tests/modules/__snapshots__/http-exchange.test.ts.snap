// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`httpExchange logs a warning when used for subscriptions 1`] = `"Received a subscription operation in the httpExchange. You are probably trying to create a subscription. Have you added a subscriptionExchange?"`;
